### PR TITLE
Fix leaks!

### PR DESCRIPTION
**Pull request by nyalldawson**
More leak fixes, some serious, some not.

### CODE REVIEW

File: src/core/geometry/qgslinestring.cpp [hunk 1]
```diff
@@ -786,24 +786,23 @@ int QgsLineString::dimension() const
...
+  std::unique_ptr< double[] > dummyZ;
+  if ( !hasZ || !transformZ )
   {
-    zArray = new double[nPoints];
-    for ( int i = 0; i < nPoints; ++i )
-    {
-      zArray[i] = 0;
-    }
+    dummyZ.reset( new double[nPoints]() );
+    zArray = dummyZ.get();
```
Comment (m-kuhn) [inline src/core/geometry/qgslinestring.cpp]: Is the 0 initialisation not needed here?

File: src/core/qgsvectorfilewriter.cpp [hunk 2]
```diff
@@ -674,1211 +674,1243 @@ OGRGeometryH QgsVectorFileWriter::createEmptyGeometry( QgsWkbTypes::Type wkbType
...
+                               QObject::tr( "Generic Mapping Tools [GMT]" ),
+                               QStringLiteral( "*.gmt" ),
+                               QStringLiteral( "gmt" ),
+                               datasetOptions,
+                               layerOptions
+                             )
+                           );
+
+      // GPS eXchange Format [GPX]
+
```
Comment (m-kuhn) [inline src/core/qgsvectorfilewriter.cpp]: Can't we just put all the init code into the constructor and skip this conditional initialisation?

File: src/core/geometry/qgslinestring.cpp (hunk 1 above)
Comment (nyalldawson) [reply src/core/geometry/qgslinestring.cpp]: It's still there - the () does it

File: src/core/qgsvectorfilewriter.cpp (hunk 2 above)
Comment (nyalldawson) [reply src/core/qgsvectorfilewriter.cpp]: Yep, good approach. This was a iterative solution :)

Comment (nyalldawson) [reply src/core/qgsvectorfilewriter.cpp]: Fixed

**Comment by m-kuhn** [comment]
Thanks for this. Merge ahead.

**Pull request merged**